src/multi_auth/provider.cr

```diff
@@ -1,11 +1,24 @@
-abstract class MultiAuth::Provider
-  getter redirect_uri : String
-  getter client_id : String
-  getter client_secret : String
+module MultiAuth
+  abstract class Provider
+    abstract def authorize_uri(scope = nil)
+    abstract def user(params : Hash(String, String))
+  end
+
+  abstract class OAuthProvider < Provider
```

Comment:
The API of both `OAuthProvider`, `OAuth2Provider ` provide is same, is not it? Shell we combine them.

Reply:
Hm, maybe. But the API is not the same:

```
OAuthProvider#oauth_callback
OAuthProvider#consumer_key
OAuthProvider#consumer_secret

OAuth2Provider#redirect_uri
OAuth2Provider#client_id
OAuth2Provider#client_secret
```

I don't like the idea to combine those attributes into one class, so each OAuth2 provider will have redundant `comsumer_key` and `consumer_secret` attributes and vice versa...

Maybe it is possible to just do:

```
Provider#callback
Provider#key
Provider#secret
```

but that may lead to confusion with the required attributes to pass in...

What is your suggestion?

Reply:
Personally, I would `just do`:

```
Provider#callback_uri
Provider#key
Provider#secret
```
All providers need three params, this described in the abstract provider.